File: week6/old-examples/Chapter 7/B04170_07_Python_Draft_01/chapter7/general/floor_plan_element.py
__author__ = 'gaston'


class FloorPlanElement:
    category = "Undefined"
    description = "Undefined"

    def __init__(self, x, y, width, height, parent):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.parent = parent

    def move_to(self, x, y):
        self.x = x
        self.y = y

    def print_category(self):
        print(type(self).category)

    def print_description(self):
        print(type(self).description)

    def draw(self):
        self.print_category()
        self.print_description()
        print("X: " + str(self.x) +
              ", Y: " + str(self.y) +
              ". Width: " + str(self.width) +
              ", Height: " + str(self.height) + ".")

File: week6/old-examples/Chapter 7/B04170_07_CSharp_Draft_01/General/FloorPlanElement.cs
namespace Chapter7.General
{
    using System;

    public abstract class FloorPlanElement: IFloorPlanElement
    {
        public virtual string Category 
        { 
            get { return "Undefined"; }
            set { throw new InvalidOperationException(); }
        }

        public virtual string Description
        {
            get { return "Undefined"; }
            set { throw new InvalidOperationException(); }
        }

        public double X { get; set; }
        public double Y { get; set; }
        public double Width { get; set; }
        public double Height { get; set; }

        private IFloorPlanElement _parent;
        public IFloorPlanElement Parent
        {
            get { return _parent; }
            set { throw new InvalidOperationException(); }
        }

        public FloorPlanElement(double x, double y, double width, double height, IFloorPlanElement parent)
        {
            this.X = x;
            this.Y = y;
            this.Width = width;
            this.Height = height;
            this._parent = parent;
        }

        public void MoveTo(double x, double y)
        {
            this.X = x;
            this.Y = y;
        }

        public void PrintCategory()
        {
            Console.WriteLine(this.Category);
        }

        public void PrintDescription()
        {
            Console.WriteLine(this.Description);
        }

        public void Draw()
        {
            this.PrintCategory();
            this.PrintDescription();
            Console.WriteLine(
                "X: {0}, Y: {1}. Width: {2}, Height: {3}.",
                this.X,
                this.Y,
                this.Width,
                this.Height);
        }
    }
}
